_source/_posts/2025-11-04-user-entitlements-workshop.md

```diff
@@ -0,0 +1,1298 @@
+
+  const roles = await prisma.role.findMany({
+    take: recordLimit,
+    skip: startIndex - 1
+  });
+
+const listResponse: IListResponse = {
+    schemas: [SCHEMA_LIST_RESPONSE],
+    totalResults: roles.length,
+    startIndex,
+    itemsPerPage: recordLimit,
+    Resources: roles.map(role => ({
+      schemas: [SCHEMA_OKTA_ROLE],
+      id: role.id.toString(),
+      displayName: role.name
+    }))
+  };
+
+  return res.json(listResponse);
+});
+```
+
+Run a quick check to ensure everything still works. Serve the API and call the `/Roles` endpoint using your HTTP client. ✅
+
+You should see three roles matching the roles in the database. 🎉
+
+## SCIM resource types 
+
+We implemented the `/Roles` endpoint and discussed how SCIM defines a resource. But how would the SCIM client know about this Okta Role type? Enter discovery—learning about a SCIM server's capabilities and supported objects such as resources!
+
+SCIM clients and servers communicate about the types of resources through a standard endpoint, the`/ResourceType` endpoint. SCIM clients call the endpoint to discover what resources they can expect. The endpoint returns a SCIM list response outlining resources. You can add every resource type used, including the standard `User` and `EnterpriseUser` resources, but Okta expects resource definitions only for custom types. 
+
+First, we'll create the interface for the `ResourceType` and define some strings. Open `okta-enterprise-ready-workshops/apps/api/src/scim-types.ts`. Add the interface for `IResourceType` above the `IListResponse` interface.
+
+
+```ts
+export interface IResourceType {
+  id?: string;
+  schemas: string[];
+  name: string; 
+  description?: string;
+  endpoint: string;
+  schema: string; 
+  meta: IMetadata;
+}
+```
+
+Notice the `IResourceType` doesn't extend from the `IScimResource` interface. For example, the SCIM standard doesn't require `id` for a resource type. Since the SCIM standard treats `ResourceType` as an exception case of `Resource`, we defined it separately without the relation instead of extending from `IScimResource`.
+
+When following the SCIM protocol, responses that list values, such as the list of roles or resource types, use the SCIM list response format.
+
+The `IListResource` interface must support `IOktaRole` and `IResourceType`. Using [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html) and [union types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types), we can support different list response objects . Update the `IListResource` to match the code below.
+
+```ts
+export interface IListResponse<T extends IScimResource | IResourceType> {
+  schemas: string[];
+  totalResults: number;
+  startIndex: number;
+  itemsPerPage: number;
+  Resources: T[];
+}
+```
+
+You'll see errors in the IDE and, if you're running the API, within the console output. No worries; we'll fix those errors soon!
```

Comment:
somewhere around here, an error showed up where prisma suddenly forgot that we'd just taught it about roles: 
<img width="947" height="120" alt="image" src="https://github.com/user-attachments/assets/ee438e43-c259-4102-afcc-83e18c1c9171" />


It went away when I ran `TypeScript: Restart TS Server` in the vscode command palette. This was probably just a random problem with my vscode; only if others end up troubleshooting the same thing would it be worth mentioning in the post.